storage/replica.go

```diff
@@ -1151,6 +1156,97 @@ func (r *Replica) maybeSetCorrupt(err error) error {
+	trace := tracer.FromCtx(ctx)
+	tracer.ToCtx(ctx, nil) // we're doing async stuff below; those need new traces
+	trace.Event("resolving intents [async]")
+	var wg sync.WaitGroup
+
+	bArgs := &proto.BatchRequest{}
+	bArgs.User = security.RootUser
+	for i := range intents {
+		intent := intents[i] // avoids a race in `i, intent := range ...`
+		var resolveArgs proto.Request
+		var local bool // whether this intent lives on this Range
+		{
+			header := proto.RequestHeader{
+				// Use the pushee's timestamp, which might be lower than the
+				// pusher's request timestamp. No need to push the intent higher
+				// than the pushee's txn!
+				Timestamp: intent.Txn.Timestamp,
+				Key:       intent.Key,
+				EndKey:    intent.EndKey,
+				User:      security.RootUser,
+				Txn:       &intent.Txn,
+			}
+
+			if len(intent.EndKey) == 0 {
+				resolveArgs = &proto.ResolveIntentRequest{RequestHeader: header}
+				local = r.ContainsKey(intent.Key)
+			} else {
+				resolveArgs = &proto.ResolveIntentRangeRequest{RequestHeader: header}
+				local = r.ContainsKeyRange(intent.Key, intent.EndKey)
+			}
+		}
+
+		// If the intent isn't (completely) local, we'll need to send an external request.
+		// We'll batch them all up and send at the end.
+		if !local {
+			bArgs.Add(resolveArgs)
+			continue
+		}
+
+		// If it is local, it goes directly into Raft.
+		wg.Add(1)
+		action := func() {
+			// Trace this under the ID of the intent owner.
+			ctx := tracer.ToCtx(ctx, r.rm.Tracer().NewTrace(resolveArgs.Header().Txn))
+			if _, err := r.addWriteCmd(ctx, resolveArgs, &wg); err != nil && log.V(0) {
+				log.Warningc(ctx, "resolve for key %s failed: %s", intent.Key, err)
+			}
+		}
+		if !r.rm.Stopper().RunAsyncTask(action) {
+			// Still run the task. Our caller already has a task and going async
+			// here again is merely for performance, but some intents need to
+			// be resolved because they might block other tasks. See #1684.
+			// Note that handleSkippedIntents has a TODO in case #1684 comes
+			// back.
+			action()
+		}
+	}
+	// Resolve all of the intents which aren't local to the Range. This is a
+	// no-op if all are local.
+	b := &client.Batch{}
+	b.InternalAddCall(proto.Call{Args: bArgs, Reply: &proto.BatchResponse{}})
+	action := func() {
+		// TODO(tschottdorf): no tracing here yet. Probably useful at some point,
+		// but needs a) the corresponding interface and b) facilities for tracing
+		// multiple tracees at the same time (batch full of possibly individual
+		// txns).
+		if err := r.rm.DB().Run(b); err != nil {
+			if log.V(1) {
+				log.Infoc(ctx, "%s", err)
+			}
+		}
+	}
+	if !r.rm.Stopper().RunAsyncTask(action) {
+		// As with local intents, try async to not keep the caller waiting, but
+		// when draining just go ahead and do it synchronously. See #1684.
+		action()
+	}
+
+	// Wait until all the local `ResolveIntent`s have been submitted to raft.
+	// No-op if all were external.
+	wg.Wait()
```

Comment:
local intents can be conflict resolution with the client waiting, so we want to wait for the minimum amount of time that guarantees that after returning to the client, they'll not hit the intent again. That's why we wait for propose and then return. BTW, nothing has changed for those semantics.

with non-local intents, we know there's no client waiting, so there's no point in blocking the caller. That's the new part introduced.

I'm updating the comments to explain that better.